### ISSUE TITLE
RStudio consumes all available RAM when rendering R Markdown causing Linux Freeze

### ISSUE THREAD

**Issue by statsandthings** [rstudio/rstudio#2319]
RStudio Version : 1.1.423
    OS Version : Ubuntu 16.04
    R Version : 3.4.3

### Steps to reproduce the problem
I cycle between the two processes shown below for time periods up to an hour. Within that hour RStudio will freeze my entire Ubuntu system and consume all available system RAM after issuing a `rmarkdown::render("rmarkdown_file.Rmd")` command. At this point my system in unusable. The two processes are:
(1) Edit my rmarkdown_file.Rmd
(2) Render it for viewing with `rmarkdown::render("rmarkdown_file.Rmd")`

### Describe the problem in detail
I have about a dozen scripts each between 100 and 1,000 lines. I run these scripts in series on several text files that are around 10 MiB in size, at a max around 50,000 rows x 20 columns. I'm building an R Markdown file that pulls in a lot of tables, plots, and analysis from these scripts and displays them as an HTML file. I spend several minutes adding to my R Markdown document, then I want to see how the changes look in the HTML file. I issue the command `rmarkdown::render("directory/your_doc.Rmd")` and 19 out of 20 times it works OK.

I've kept my system monitor up and typically rstudio consumes 150 MiB of RAM. rsession consumes around 200 MiB of ram. Each time this issue occurs (the 20th out of 20 times), it's in the following sequence: 

- spend several minutes editing Rmd file
- Save Rmd file
- issue rmarkdown::render("rmarkdown_file.Rmd") command
- within seconds Ubuntu freezes but I can still hazardly move the mouse around
- frantically click on System Monitor icon for several minutes until it randomly responds
- System Monitor shows rsession consuming 7+ GiB RAM
- I only have 8 GiB RAM and system is almost totally frozen, can't do anything
- hard reset

### Describe the behavior you expected
No frozen Ubuntu session causing a hard restart every 30-60 minutes. For note the system never truly crashes per se, rsession just consumes all available RAM and you can't do anything at all. I even try to drop back to the terminal <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>F1</kbd> and that won't work. The system is totally fronzen for several minutes and then for a few seconds it's barely responsive and you can sneak stuff in.

Thank you for for looking into this. If you need more info from me just ask. I should also mention the same behavior was happening in R Studio 1.1.383.

**Comment by kevinushey** [comment]
Can you please provide a reproducible example? If at all possible, a (standalone) R Markdown document that we can try to render to reproduce the issue on our own machines.

**Comment by statsandthings** [comment]
I'm trying to create a bare bones reproducible and am having trouble reproducing the issue. I can't submit my original data set which is sensitive information. I'll continue to try to get a reproducible I can post. I switched over to Windows 10 from Ubuntu 16.04 (just for troubleshooting) - maybe it was Linux related. Nope. 

I started working on the same R Markdown file on Windows and this time I get a weird (to me) error instead of the entire system freeze. Again, 19 out of 20 times I'm able to render the R Markdown into an HTML file that's around 700 KB (kilobytes). The 20th time I try I get `Error: cannot allocate vector of size 17.6 Gb`. BUT I only added a few paragraphs to the Rmd file and some math equations?? How can it go from 700 KB to 17.6 Gb? Here's the complete output if it helps.

What is *running away* in this process?

```
rmarkdown::render("mark_output.Rmd")


processing file: mark_output.Rmd
  |.........                                                        |  14%
  ordinary text without R code

  |...................                                              |  29%
label: setup (with options) 
List of 1
 $ include: logi FALSE

  |............................                                     |  43%
  ordinary text without R code

  |.....................................                            |  57%
label: cars
  |..............................................                   |  71%
  ordinary text without R code

  |........................................................         |  86%
label: pressure (with options) 
List of 3
 $ fig.height: num 7
 $ fig.width : language 12 + (4/9)
 $ echo      : logi FALSE

  |.................................................................| 100%
  ordinary text without R code


output file: mark_output.knit.md

"C:/Users/statsandthings/AppData/Local/Pandoc/pandoc" +RTS -K512m -RTS mark_output.utf8.md --to html --from markdown+autolink_bare_uris+ascii_identifiers+tex_math_single_backslash+smart --output mark_output.html --email-obfuscation none --self-contained --standalone --section-divs --template "C:\Users\10258753\Documents\Rlibs\rmarkdown\rmd\h\default.html" --no-highlight --variable highlightjs=1 --variable "theme:bootstrap" --include-in-header "C:\Users\10258753\AppData\Local\Temp\RtmpGi2lRl\rmarkdown-str460774e669e.html" --mathjax --variable "mathjax-url:https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML" 

Output created: mark_output.html
Error: cannot allocate vector of size 17.6 Gb
```

**Comment by kevinushey** [comment]
That's bizarre! But I still don't have a good idea what could be causing this :-/

**Comment by AlexanderSeverin** [comment]
I am having a similar issue

Ubuntu 16.04.3 (Amazon EC2)
R version 3.2.3
R Studio server 1.0.143

I render some files, everything works, then I render the same file, nothing changed except the parameter in Rmd-file. Suddenly all out of memory and have to reboot. Then sometimes no need for a reboot, but get this error message 

Quitting from lines 424-442 (slag_coop.Rmd) 
 Error in 1:max(year_sums_week$week) : result would be too long a vector In addition: Warning messages:
1: In styling_latex_scale_down(out, table_info) :

Checking memory at this point shows almost zero available.

**Comment by kevinushey** [comment]
Reproducible example?

**Comment by jmcphers** [comment]
> (1) Edit my rmarkdown_file.Rmd
> (2) Render it for viewing with rmarkdown::render("rmarkdown_file.Rmd")

Another thing you might try is doing both of these things outside RStudio (i.e. edit with a plain editor and render using R in a terminal or via RGui). Error messages such as `Error: cannot allocate vector of size 17.6 Gb` are from R rather than RStudio. Since this happens during `rmarkdown::render`, there's a good chance that something in one of the R code chunks in your document or in the rmarkdown package is triggering the runaway resource usage.

**Comment by areedv** [comment]
I'm having a similar symptom running ```rmarkdown::render``` in an app hosted by shiny-server: after a few runs memory usage accumulates until R complains with ```Warning: Error in : cannot allocate vector of size 54.1 Gb```. Eventually, shiny-server have to be restarted.

Calling garbage collection (```gc()```) immediately after ```rmarkdown::render``` seems to have an effect on this: so far there has been no need for a shiny-server restart and memory usage seems to peak rather than accumulate (even though memory usage can be quite extensive when rendering from Rmd-documents).

I just recently applied garbage collection to my R-code so I am not able to conclude if this will be a viable solution in my case. Also, this is somewhat surprising since I regarded garbage collection in R more of an implicit feature

**Comment by nikkoc** [comment]
I've been having an issue that sounds just like this. My Shiny app generates a PDF report from an Rmd file as part of a download handler. The PDF that I test with is about 7 pages. By testing the same download button in the same session, the first 4-5 PDF renders go smoothly. However, after the nth iteration, the rsession memory jumps sizably (to ~ 1GB, up from ~250MB average), then n+1 iteration the memory jumps into the 7-8 GB range and climbs. 

After stepping the code inside `render`, I found that [old_knit_meta <- knit_meta_reset()](https://github.com/rstudio/rmarkdown/blob/bbd0786d82f82d9eaef7f0b9e2c553b523407161/R/render.R#L417) was somehow not getting cleared. For example, whereas it should typically register around 40 bytes for an empty list, it would register 843317640 bytes. `knit_meta_reset()` calls `knitr::knit_meta()`, which defaults to `clean=T`, so it should clear the `.knitEnv$meta` after it returns it. It usually does, but for some weird reason that I don't know, it doesn't in this circumstance and thus causes old_knit_meta to get really large. 

What I found was that if I 'reset' the knit_meta myself by calling `knitr::knit_meta(clean=T)` when debugging the code in the ~1GB memory iteration, it would release that memory and go back to the ~250MB range (following @areedv 's comment, I call `gc()` after `render`). As a result I now 'force' knit_meta to be empty by calling `knit_meta` before I call `rmarkdown::render`. I made a function specifically to wrap the `knit_meta` call inside `invisible()` to avoid printing the results.

I have only done this today, so I haven't run into the issue YET. But when I tested this solution, I grew tired of running the same PDF after about 25 times and I did not see any anomalous spike in memory consistent with this bug.
